File: compress_logo.py
import os
import sys

# Try to import Pillow. If not present, try to install it.
try:
    from PIL import Image
except ImportError:
    print("Pillow is not installed. Installing it via pip...")
    import subprocess
    subprocess.check_call([sys.executable, "-m", "pip", "install", "Pillow"])
    from PIL import Image

def compress_logo():
    input_path = os.path.join("assets", "icon.png")
    output_path = os.path.join("assets", "icon_compressed.png")
    
    if not os.path.exists(input_path):
        print(f"Error: Could not find {input_path}")
        return
        
    original_size = os.path.getsize(input_path) / (1024 * 1024)
    print(f"Original logo file size: {original_size:.2f} MB")
    
    print("Opening and processing the image...")
    with Image.open(input_path) as img:
        print(f"Original dimensions: {img.width}x{img.height}")
        
        # Determine new size (max width/height of 512px)
        max_size = 512
        if img.width > max_size or img.height > max_size:
            if img.width > img.height:
                new_width = max_size
                new_height = int(img.height * (max_size / img.width))
            else:
                new_height = max_size
                new_width = int(img.width * (max_size / img.height))
            
            print(f"Resizing to: {new_width}x{new_height}")
            # Use high-quality Resampling.LANCZOS (or ANTIALIAS fallback)
            try:
                resample_method = Image.Resampling.LANCZOS
            except AttributeError:
                resample_method = Image.ANTIALIAS
                
            img_resized = img.resize((new_width, new_height), resample_method)
        else:
            print("Image is already small, keeping original dimensions.")
            img_resized = img
            
        print("Saving optimized version...")
        # Save as PNG with optimal compression
        img_resized.save(output_path, "PNG", optimize=True)
        
    compressed_size = os.path.getsize(output_path) / 1024
    print(f"Compressed logo file size: {compressed_size:.2f} KB")
    
    # Overwrite the original logo with the optimized one
    os.replace(output_path, input_path)
    print("✓ Success! Overwrote original assets/icon.png with optimized web version.")

if __name__ == "__main__":
    compress_logo()

File: compress_logo.php
<?php
/**
 * Script to compress and resize the 39MB icon.png file to a web-optimized size.
 */

$input = 'assets/icon.png';
$output = 'assets/icon_compressed.png';

if (!file_exists($input)) {
    die("Error: assets/icon.png not found.\n");
}

$size_mb = filesize($input) / (1024 * 1024);
echo "Original size: " . round($size_mb, 2) . " MB\n";

if (!extension_loaded('gd')) {
    die("GD extension is not loaded. Please make sure GD is enabled in your PHP installation.\n");
}

echo "Opening PNG image (this might take a few seconds because of its 39MB size)...\n";
// Increase memory limit for this process since 39MB image needs lots of RAM to decode
ini_set('memory_limit', '512M');

$src = @imagecreatefrompng($input);
if (!$src) {
    die("Error: Could not open the PNG image. It might be corrupt or too large for the current memory limit.\n");
}

$width = imagesx($src);
$height = imagesy($src);
echo "Original dimensions: {$width}x{$height}\n";

$max_size = 512;
if ($width > $max_size || $height > $max_size) {
    if ($width > $height) {
        $new_width = $max_size;
        $new_height = intval($height * ($max_size / $width));
    } else {
        $new_height = $max_size;
        $new_width = intval($width * ($max_size / $height));
    }
} else {
    $new_width = $width;
    $new_height = $height;
}

echo "Resizing to {$new_width}x{$new_height}...\n";
$dst = imagecreatetruecolor($new_width, $new_height);

// Maintain transparency for PNG
imagealphablending($dst, false);
imagesavealpha($dst, true);
$transparent = imagecolorallocatealpha($dst, 255, 255, 255, 127);
imagefill($dst, 0, 0, $transparent);

imagecopyresampled($dst, $src, 0, 0, 0, 0, $new_width, $new_height, $width, $height);

echo "Saving optimized PNG...\n";
// 9 is maximum compression
if (imagepng($dst, $output, 9)) {
    imagedestroy($src);
    imagedestroy($dst);

    $new_size_kb = filesize($output) / 1024;
    echo "Optimized size: " . round($new_size_kb, 2) . " KB\n";

    if (rename($output, $input)) {
        echo "✓ Success! Overwrote assets/icon.png with optimized version.\n";
    } else {
        echo "Error: Could not replace the original file.\n";
    }
} else {
    echo "Error: Failed to save the optimized image.\n";
}
